src/doc_iter.rs

```diff
@@ -0,0 +1,301 @@
+    Name(&'a str),
+    Index(usize),
+}
+
+impl std::fmt::Display for DocPath<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            DocPath::Name(n) => write!(f, "{}", n),
+            DocPath::Index(i) => write!(f, "{}", i),
+        }
+    }
+}
+
+impl<'a> Iterator for DocPathIter<'a> {
+    type Item = (Vec<DocPath<'a>>, &'a Document);
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(top) = self.stack.last_mut() {
+            if let Some(val) = top.next() {
+                match val {
+                    Document::Comment(_, _) => continue,
+                    Document::Mapping(v) => {
+                        self.stack.push(v.iter());
+                        self.path.push(DocPath::Name(""));
+                        self.aggregate.push(true);
+                        continue;
+                    }
+                    Document::Sequence(v) => {
+                        self.stack.push(v.iter());
+                        self.path.push(DocPath::Index(usize::MAX));
+                        self.aggregate.push(true);
+                        continue;
+                    }
+                    Document::Compact(v) => {
+                        self.stack.push(std::slice::from_ref(&**v).iter());
+                        self.aggregate.push(false);
+                        continue;
+                    }
+                    Document::Fragment(f) => {
+                        match self.path.last_mut() {
+                            Some(DocPath::Name(ref mut n)) => match val.as_kv() {
+                                Ok((k, v)) => {
+                                    *n = k.as_str().expect("DocPath key");
+                                    self.stack.push(std::slice::from_ref(v).iter());
+                                }
+                                Err(_) => continue,
+                            },
+                            Some(DocPath::Index(_)) => match val.as_value() {
+                                Ok(v) => self.stack.push(std::slice::from_ref(v).iter()),
+                                Err(_) => continue,
+                            },
+                            _ => {
+                                self.stack.push(f.iter());
+                            }
+                        };
+                        self.aggregate.push(false);
+                        continue;
+                    }
+                    _ => {
+                        if let Some(DocPath::Index(ref mut i)) = self.path.last_mut() {
+                            *i = i.wrapping_add(1);
+                        }
+                        return Some((self.path.clone(), val));
+                    }
+                }
+            } else {
+                self.stack.pop();
+                if self.aggregate.pop() == Some(true) {
+                    self.path.pop();
+                }
+            }
+        }
+        None
+    }
```

Comment:
As you noted, the `continue` statements here prevent you from breaking this function into two sequential steps like we did with `DocIter::next`. However, you can still remove one layer of nesting here, by transforming the following:

```rust
if let Some(val) = top.next() {
    /* match statement */
} else {
    self.stack.pop();
    if self.aggregate.pop() == Some(true) {
        self.path.pop();
    }
}
```

into:

```rust
let val = if let Some(val) = top.next() {
    val
} else {
    self.stack.pop();
    if self.aggregate.pop() == Some(true) {
        self.path.pop();
    }
    continue;
};
/* match statement here */
```

This works in `DocPathIterMut::next` as well.

Reply:
I suppose this is a matter of style;  **to me**, your example propagates `val` to beyond the scope where it should live and complicates the flow control by putting the `continue` statement into the `else` arm of the conditional `if`-expr.

Another possible rewrite would be:
```
match top.next() {
    Some(Document::Comment(_,_)) => { ... }
    Some(Document::Mapping(v)) => { ... }
    ...
    None => {
         self.stack.pop();
         if self.aggregate.pop() == Some(true) { self.path.pop(); }
    }
}
```
Again, I'm not certain that this is better (also for style reasons).  I prefer destructuring the `Some(...)` before the match expression.

Reply:
Yes, it is a matter of style. I'm mainly trying to decrease the amount of nesting in this function. If you think it makes more sense to keep the nesting in place, then that's fine with me.

Reply:
Ok, I've gone with the `Some(Document::xxx))` patterns in the match expression to eliminate the inner `if let` expression.  Once I wrote the whole thing out, it didn't look as bad as I thought it would.